src/Lean/Compiler/LCNF/FVarUtil.lean

```diff
@@ -0,0 +1,156 @@
+/-
+Copyright (c) 2022 Henrik Böving. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Henrik Böving
+-/
+import Lean.Expr
+import Lean.Compiler.LCNF.Basic
+import Lean.Compiler.LCNF.CompilerM
+
+namespace Lean.Compiler.LCNF
+
+class TraverseFVar (α : Type) where
+  mapFVarM {m : Type → Type} [MonadLiftT CompilerM m] [Monad m] (f : FVarId → m FVarId) (val : α) : m α
+  forFVarM {m : Type → Type} [Monad m] (f : FVarId → m Unit) (val : α) : m Unit
+
+export TraverseFVar (mapFVarM forFVarM)
+
+partial def Expr.mapFVarM [MonadLiftT CompilerM m] [Monad m] (f : FVarId → m FVarId) (e : Expr) : m Expr := do
+  match e with
+  | .proj typ idx struct => return .proj typ idx (← mapFVarM f struct)
+  | .app fn arg => return .app (← mapFVarM f fn) (← mapFVarM f arg)
+  | .fvar fvarId => return .fvar (← f fvarId)
+  | .lam arg ty body bi =>
+    return .lam arg (← mapFVarM f ty) (← mapFVarM f body) bi
+  | .forallE arg ty body bi =>
+    return .forallE arg (←mapFVarM f ty) (← mapFVarM f body) bi
+  | .letE var ty value body nonDep  =>
+    return .letE var (← mapFVarM f ty) (← mapFVarM f value) (← mapFVarM f body) nonDep
+  | .bvar .. | .sort .. => return e
+  | .mdata .. | .const .. | .lit .. => return e
+  | .mvar .. => unreachable!
+
+partial def Expr.forFVarM [Monad m] (f : FVarId → m Unit) (e : Expr) : m Unit := do
+  match e with
+  | .proj _ _ struct => forFVarM f struct
+  | .app fn arg =>
+    forFVarM f fn
+    forFVarM f arg
+  | .fvar fvarId => f fvarId
+  | .lam _ ty body .. =>
+    forFVarM f ty
+    forFVarM f body
+  | .forallE _ ty body .. =>
+    forFVarM f ty
+    forFVarM f body
+  | .letE _ ty value body ..  =>
+    forFVarM f ty
+    forFVarM f value
+    forFVarM f body
+  | .bvar .. | .sort .. => return
+  | .mdata .. | .const .. | .lit .. => return
+  | .mvar .. => unreachable!
+
+instance : TraverseFVar Expr where
+  mapFVarM := Expr.mapFVarM
+  forFVarM := Expr.forFVarM
+
+partial def LetDecl.mapFVarM [MonadLiftT CompilerM m] [Monad m] (f : FVarId → m FVarId) (decl : LetDecl) : m LetDecl := do
+  decl.update (← Expr.mapFVarM f decl.type) (← Expr.mapFVarM f decl.value)
+
+partial def LetDecl.forFVarM [Monad m] (f : FVarId → m Unit) (decl : LetDecl) : m Unit := do
+  Expr.forFVarM f decl.type
+  Expr.forFVarM f decl.value
+
+instance : TraverseFVar LetDecl where
+  mapFVarM := LetDecl.mapFVarM
+  forFVarM := LetDecl.forFVarM
+
+partial def Code.mapFVarM [MonadLiftT CompilerM m] [Monad m] (f : FVarId → m FVarId) (c : Code) : m Code := do
+  match c with
+  | .let decl k =>
+    let decl ← LetDecl.mapFVarM f decl
+    return Code.updateLet! c decl (← mapFVarM f k)
+  | .fun decl k =>
+    -- XXX: We do not visit params...is this okay/intuitive?
```

Comment:
We should visit params.

Reply:
To explain my thought process here. I do also not visit e.g. the FVarId's that are names of say the let declarations but only the ones that are actually mentioned in the code so my thought process was to treat the parameters similar since they should occur in values or types if they are being used.

If I am supposed to visit parameters, should I also visit the FVarIds of the declarations?

Reply:
I meant visiting the parameter types, they may contain free variable ids.